### PR TITLE
Apply apply_group_norm_silu to LTX-2 latent upsampler

### PR DESCRIPTION
**Pull request by pengdurice**
## Motivation

The LTX-2 spatial **`LatentUpsampler`** (`runtime/models/upsampler/latent_upsampler.py`) is auto-loaded as the `spatial_upsampler` component of every LTX-2 two-stage pipeline — Lightricks generates latents at lower spatial resolution and uses this module to upscale them by 2× before VAE decode, so it sits on the critical path of every LTX-2 generation. Its `LatentUpsampler` and `ResBlock` modules use `torch.nn.GroupNorm(32, ...) + torch.nn.SiLU()` at multiple sites, but the existing inline `silu(norm(x))` pattern bypasses the Triton-fused `apply_group_norm_silu` helper that HunyuanVAE already routes through.

This PR plumbs the helper into the three eligible direct-`silu(norm(x))` sites in `latent_upsampler.py`, picking up a **6.5×–14× kernel speedup on H100** (vs eager `silu(group_norm(x))`) without changing model outputs. The speedup factors into a ~1.55× SiLU+GroupNorm fusion win (constant across shapes) and a 4×–9× parallelization win that scales with per-group size; see [Speed Tests and Profiling](#speed-tests-and-profiling) for the decomposition.

This also addresses the "kernel fusion instead of compiler-driven fusion" line under Performance Improvements in [Roadmap] SGLang-Diffusion (26 Q2) (#23035).

## Modifications

- **`python/sglang/multimodal_gen/runtime/models/upsampler/latent_upsampler.py`** — route three `silu(norm(x))` sites through `apply_group_norm_silu`:

  | Site | Class | Hit per `LatentUpsampler.forward` |
  |---|---|---|
  | `initial_norm` + `initial_activation` | `LatentUpsampler.forward` (both 2D and 3D paths) | 1 |
  | `norm1` + `activation` | `ResBlock.forward` (one per ResBlock) | `2 * num_blocks_per_stage` |

  At the default `num_blocks_per_stage=4` this is **9 fused sites per upsampler call** (1 initial + 4 pre-upsample ResBlocks + 4 post-upsample ResBlocks).

  The second norm site inside `ResBlock` (`silu(norm2(x) + residual)` at line 118) is deliberately left on the eager path with an inline comment — the current `apply_group_norm_silu` helper only covers `silu(norm(x))`, not the `silu(norm(x) + residual)` pattern. A follow-up `apply_group_norm_add_silu` kernel would catch those remaining 8 sites per generation (doubling fusion coverage on this module); it's out of scope here.

- **`python/sglang/jit_kernel/benchmark/diffusion/bench_group_norm_silu.py`** — add three LTX-2 latent upsampler shape cases (`ltx2_upsampler_small`, `ltx2_upsampler_pre_720p`, `ltx2_upsampler_post_720p`) at `num_groups=32`, bracketing the production shape range that Lightricks ships in the LTX-2.3 spatial upscaler checkpoints. The `post_720p` case is ~471M bf16 elements (~940 MB), making the kernel asymptotically HBM-bandwidth-bound. This bench is registered with `disabled="standalone benchmark"`, so it never appears in the regular CI memory budget.

- **`python/sglang/multimodal_gen/test/unit/test_latent_upsampler_group_norm_silu.py`** (new file) — 19 unit tests (10 CPU-fp32 wiring + 9 CUDA bf16/fp16 production-path parity, the CUDA tests gated on `torch.cuda.is_available()`).

## Accuracy Tests

Zero-tolerance parity vs the original eager `silu(norm(x))` chain on CPU plus CUDA-tolerance parity on the production fast path:

```bash
$ PYTHONPATH=python CUDA_VISIBLE_DEVICES=0 .venv/bin/python -m pytest -p no:cacheprovider \
    python/sglang/multimodal_gen/test/unit/test_latent_upsampler_group_norm_silu.py -v
...
================ 19 passed, 16 warnings in 8.89s ================
```

**Coverage:**

| Test | Tier | What it pins |
|---|---|---|
| `test_resblock_forward_parity` (3 parametrized) | CPU fp32 | `ResBlock.forward` output bit-identical (`atol=0.0, rtol=0.0`) to an explicit `activation(norm(x))` reference. CPU helper fallback path. |
| `test_latent_upsampler_forward_parity` (2 parametrized) | CPU fp32 | Same for `LatentUpsampler.forward` across 2D + 3D conv paths. |
| `test_resblock_fuses_exactly_one_site` | CPU fp32 | Patches `apply_group_norm_silu` and asserts exactly 1 call per `ResBlock.forward` (the residual-add `norm2` site must stay eager). |
| `test_latent_upsampler_fuses_expected_sites` (3 parametrized) | CPU fp32 | Asserts call count equals `1 + 2 * num_blocks_per_stage` (i.e., 5, 9 across configs). |
| `test_latent_upsampler_module_wires_fused_helper` | CPU fp32 | Smoke test: module imports `apply_group_norm_silu` at top level. |
| **`test_resblock_forward_parity_cuda`** (4 parametrized) | **CUDA bf16/fp16** | `ResBlock.forward` on CUDA actually fires `triton_group_norm_silu`; output matches eager reference within kernel-side tolerance (bf16: `atol=7e-2, rtol=2e-2`; fp16: `atol=3e-3, rtol=3e-3`). |
| **`test_latent_upsampler_forward_parity_cuda`** (4 parametrized) | **CUDA bf16/fp16** | Full `LatentUpsampler.forward` end-to-end parity. fp16 tolerance is intentionally looser (`atol=2e-2, rtol=1e-1`) to absorb the multi-layer fp16 quantization drift through the 8+ downstream conv layers; bf16 stays at kernel-side tolerance because bf16 has fp32-equivalent exponent range. |
| **`test_resblock_actually_uses_triton_kernel_cuda`** | **CUDA bf16** | Patches `triton_group_norm_silu` and asserts `call_count >= 1` — explicit proof the production path doesn't silently fall back to eager. |

The Triton kernel itself is already covered by `python/sglang/jit_kernel/tests/diffusion/test_group_norm_silu.py` (parametrized over `(shape, num_groups, dtype)`, registered via `register_cuda_ci`).

## Speed Tests and Profiling

Measured on H100 SXM5 (`p5.48xlarge`, `CUDA_VISIBLE_DEVICES=0`) via the bench script added in this PR:

```bash
PYTHONPATH=python CUDA_VISIBLE_DEVICES=0 .venv/bin/python \
    python/sglang/jit_kernel/benchmark/diffusion/bench_group_norm_silu.py \
    --cases ltx2_upsampler_small,ltx2_upsampler_pre_720p,ltx2_upsampler_post_720p \
    --dtypes bf16,fp16 --rounds 3 --warmup 25 --rep 100
```

| case | dtype | shape | native (µs) | fused (µs) | speedup |
|---|---|---|---:|---:|---:|
| `ltx2_upsampler_small`     | bf16 | 1×512×8×45×80     |    452.00 |   70.05 |  **6.45×** |
| `ltx2_upsampler_small`     | fp16 | 1×512×8×45×80     |    451.49 |   69.22 |  6.52× |
| `ltx2_upsampler_pre_720p`  | bf16 | 1×512×16×90×160   |  3 468.27 |  261.87 | **13.24×** |
| `ltx2_upsampler_pre_720p`  | fp16 | 1×512×16×90×160   |  3 468.50 |  260.26 | 13.33× |
| `ltx2_upsampler_post_720p` | bf16 | 1×512×16×180×320  | 13 710.72 |  987.52 | **13.88×** |
| `ltx2_upsampler_post_720p` | fp16 | 1×512×16×180×320  | 13 717.54 |  984.90 | 13.93× |

Geomean kernel speedup on LTX-2 upsampler shapes: **~10.7×**.

**Where the speedup comes from.** A 13.9× speedup deserves scrutiny, so I forced the kernel onto its non-chunked one-pass path (same parallelism as PyTorch's `native_group_norm` CUDA kernel — one block per `(batch, group)` pair) and re-measured. The headline decomposes cleanly into two roughly multiplicative factors:

| case | fusion-only<br/>`native / one-pass` | parallelization<br/>`one-pass / chunked` | total<br/>`native / chunked` |
|---|---:|---:|---:|
| `ltx2_upsampler_small`     | 1.56× | 3.94× | 6.13× |
| `ltx2_upsampler_pre_720p`  | 1.56× | 8.57× | 13.34× |
| `ltx2_upsampler_post_720p` | 1.55× | 8.96× | 13.91× |

1. **Fusion contributes a flat ~1.55×**, constant across shapes — exactly what's expected for collapsing `F.silu(...)` into the apply pass: one full-tensor read + write saved per call, theoretical ceiling 2×.

2. **Chunked Welford parallelization contributes 4× → 9×**, scaling with per-group size. PyTorch's `native_group_norm` CUDA kernel (and Triton's own one-pass fallback) both launch one block per `(batch, group)` pair → **32 blocks on a 132-SM H100, ~24% SM occupancy**, with per-block work serialized across the entire `group_size` (461 K → 14.7 M elements as shape grows). The production `_launch_chunked` path splits each group into 8 192-element chunks reduced in parallel → **~57 600 programs for `post_720p`**, achieving **~2.87 TB/s = 86% of H100 HBM peak**. PyTorch's path tops out at ~12% of HBM peak on the same workload — not a memory-traffic problem, a parallelism problem.

To reproduce the decomposition, override the threshold at runtime:

```python
from sglang.jit_kernel.diffusion.triton import group_norm_silu as gns
gns._LARGE_GROUP_THRESHOLD = 10**10   # force one-pass path
```

**Where this generalizes.** Smaller DiT-side `group_norm` call sites will see only the fusion-only **~1.5×** factor; the 6×+ headline numbers are specific to **large VAE/upsampler-style shapes** where per-group size is large enough that PyTorch's per-group kernel is severely under-parallelized on Hopper. The existing HunyuanVAE wiring that this PR follows has similar (large per-group) shapes, which is why the same kernel pays off there too.

**E2E translation.** Each `LatentUpsampler.forward` call fires the helper 9 times — 5 sites at the pre-upsample resolution (1 initial + 4 pre-upsample ResBlocks) and 4 sites at the post-upsample resolution (4 post-upsample ResBlocks). For a 720p clip, naive multi-site savings:

| Sites | Per-call saved (bf16) | × count | Subtotal |
|---|---:|---:|---:|
| Pre-upsample (initial + 4 pre ResBlocks) | ~3.2 ms | × 5 | ~16 ms |
| Post-upsample (4 post ResBlocks) | ~12.7 ms | × 4 | ~51 ms |

**Total saved per generation: ~67 ms** of `LatentUpsampler.forward` wall time on H100. The latent upsampler is one bounded subsystem of a multi-second LTX-2 generation, so relative E2E gain is small in absolute terms — but it represents 20–45% of the upsampler's own runtime budget, and every LTX-2 generation pays it.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

---
### CI States

Latest PR Test (Base): :x: [Run #26735497021](https://github.com/sgl-project/sglang/actions/runs/26735497021)
Latest PR Test (Extra): :x: [Run #26735496918](https://github.com/sgl-project/sglang/actions/runs/26735496918)

### CODE REVIEW

**Comment by mickqian** [comment]
excellent. cc @Bbuf for review

File: python/sglang/multimodal_gen/test/unit/test_latent_upsampler_group_norm_silu.py [hunk 1]
```diff
@@ -0,0 +1,462 @@
+"""CPU parity tests for the LTX-2 latent upsampler GroupNorm+SiLU fusion.
```
Comment (BBuf) [inline python/sglang/multimodal_gen/test/unit/test_latent_upsampler_group_norm_silu.py]: This test file is very complex, can we simplify it?

**Review by BBuf — APPROVED** [review]

**Comment by BBuf** [comment]
Please simplify test and avoid lots of AI generated commets in test file.

**Comment by pengdurice** [comment]
@BBuf ,  thank you for the review, just simplified the test file and removed comments. thank you;-)

**Comment by BBuf** [comment]
> @BBuf , thank you for the review, just simplified the test file and removed comments. thank you;-)

Please fix lint

**Comment by BBuf** [comment]
https://github.com/sgl-project/sglang/actions/runs/26735497021/job/78788064194?pr=26045 Merged with all diffusion ci green.
